### PR TITLE
FreeBSD support in endian.h

### PR DESCRIPTION
**Pull request by mallniya**

### CODE REVIEW

File: include/dmlc/endian.h [hunk 1]
```diff
@@ -8,7 +8,7 @@
 
 #include "./base.h"
 
-#if defined(__APPLE__) || defined(_WIN32)
+#if defined(__APPLE__) || defined(__FreeBSD__) || defined(_WIN32)
```
Comment (hcho3) [inline include/dmlc/endian.h]: FreeBSD is not necessarily little endian. See https://lists.freebsd.org/pipermail/freebsd-questions/2008-February/167918.html.
To determine endianness of FreeBSD, we can include `<sys/endian.h>`

Comment (hcho3) [reply include/dmlc/endian.h]: FYI: `<sys/endian.h>` is available on all BSDs, according to this doc: http://man7.org/linux/man-pages/man3/endian.3.html

Comment (hcho3) [reply include/dmlc/endian.h]: @mallniya In the future, I will add FreeBSD to CI tests to ensure that dmlc-core compiles on FreeBSD. For now, let me update this PR to include better endian detection logic, as found in https://www.boost.org/doc/libs/1_54_0/boost/detail/endian.hpp

Comment (hcho3) [reply include/dmlc/endian.h]: @mallniya Can you try compiling with my changes?

Comment (mallniya) [reply include/dmlc/endian.h]: @hcho3 yes, compilation succeed.

Comment (hcho3) [reply include/dmlc/endian.h]: Cool!

**Comment by tqchen** [comment]
I would say let us keep endian simple, and ask users to explicit declare it

**Comment by hcho3** [comment]
@tqchen Any concrete suggestion about how to move forward with FreeBSD? Should we ask FreeBSD users to use CMake?

**Comment by tqchen** [comment]
Yes, i think use cmake is a good suggestion

**Comment by tqchen** [comment]
If we can add a quick macro support BSD, I am fine with that

**Comment by mallniya** [comment]
That's really good news, thank you!

**Pull request merged**